### PR TITLE
conv2d

### PR DESCRIPTION
**Pull request by migeed-z**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61310
* #61265
* #61239
* #61238
* #61188
* #61150
* **#61093**

Differential Revision: [D29562478](https://our.internmc.facebook.com/intern/diff/D29562478)

### CODE REVIEW

**Comment by migeed-z** [comment]
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29562478).

File: torch/fx/experimental/graph_gradual_typechecker.py [hunk 1]
```diff
@@ -198,6 +200,53 @@ def bn2d_inference_rule(n: Node, op_type):
     else:
         raise TypeError(f'Cannot apply {op_type} with input type { arg_type} and existing type {n.type} on {n}')
 
+def calculate_hout(h_in, op_type):
+    if h_in == Dyn:
+        return Dyn
+
+    elif isinstance(h_in, int):
+        h_out = floor((h_in + (2 * op_type.padding[0] - op_type.dilation[0] *
```
Comment (eellison) [inline torch/fx/experimental/graph_gradual_typechecker.py:208]: nit: i would factor out `calculate_hout` and `calculate_wout`, and parameterize by name/index, especially since Conv3d is the same exact logic for the fifth dimension.

File: torch/fx/experimental/graph_gradual_typechecker.py [hunk 2]
```diff
@@ -198,6 +200,53 @@ def bn2d_inference_rule(n: Node, op_type):
...
+    else:
+        raise TypeError(f'{w_in} in {op_type} must be a number or Dyn')
+
+@register_inference_rule(Conv2d)
+def conv2d_inference_rule(n: Node, op_type):
+    assert isinstance(n.args[0], Node)
+    n.args[0].type = apply_matching(n.args[0].type, 4)
+    arg_type = n.args[0].type
+    n.type = apply_matching(n.type, 4)
+    if is_consistent(arg_type.__args__[1], op_type.in_channels) and \
```
Comment (eellison) [inline torch/fx/experimental/graph_gradual_typechecker.py:232]: nit: i dont know that you actually need the `\` here...

File: torch/fx/experimental/graph_gradual_typechecker.py [hunk 3]
```diff
@@ -198,6 +200,53 @@ def bn2d_inference_rule(n: Node, op_type):
...
+        return w_out
+    else:
+        raise TypeError(f'{w_in} in {op_type} must be a number or Dyn')
+
+@register_inference_rule(Conv2d)
+def conv2d_inference_rule(n: Node, op_type):
+    assert isinstance(n.args[0], Node)
+    n.args[0].type = apply_matching(n.args[0].type, 4)
+    arg_type = n.args[0].type
+    n.type = apply_matching(n.type, 4)
```
Comment (eellison) [inline torch/fx/experimental/graph_gradual_typechecker.py:231]: what does this line accomplish ? since we re-assign to n.type below

File: torch/fx/experimental/graph_gradual_typechecker.py [hunk 4]
```diff
@@ -198,6 +200,53 @@ def bn2d_inference_rule(n: Node, op_type):
...
+        return h_out
+    else:
+        raise TypeError(f'{h_in} must be a number or Dyn')
+
+def calculate_wout(w_in, op_type):
+    if w_in == Dyn:
+        return Dyn
+
+    elif isinstance(w_in, int):
+        w_out = floor((w_in + (2 * op_type.padding[1] - op_type.dilation[1] *
```
Comment (eellison) [inline torch/fx/experimental/graph_gradual_typechecker.py:219]: nit: I would rewrite `floor(...)` to just use floor division `(op_type.kernel_size[1] - 1) - 1)) /` -> `(op_type.kernel_size[1] - 1) - 1)) //`

it's probably unlikely that floor(`x/y`) is different than `x // y` but sometimes floating point math is weird like that lol, and i think doing it this way is more in line with how it is computed in eager..

**Review by eellison — COMMENTED** [review]
lgtm!  just small comments

**Review by eellison — APPROVED** [review]

File: torch/fx/experimental/graph_gradual_typechecker.py [hunk 5]
```diff
@@ -207,6 +209,53 @@ def bn2d_inference_rule(n: Node, module_instance):
...
+                               (op_type.kernel_size[1] - 1) - 1)) /
+                      op_type.stride[1]) + 1
+        return w_out
+    else:
+        raise TypeError(f'{w_in} in {op_type} must be a number or Dyn')
+
+@register_inference_rule(Conv2d)
+def conv2d_inference_rule(n: Node, op_type):
```
Comment (jamesr66a) [inline torch/fx/experimental/graph_gradual_typechecker.py:236]: Similarly to the previous PR, can you rename `op_type` and add some docs?

File: test/fx/test_gradual_type.py [hunk 6]
```diff
@@ -7,6 +7,11 @@
 from torch.fx.experimental.rewriter import RewritingTracer
 from torch.fx import GraphModule
 
+def conv3x3(in_planes, out_planes, stride=1, groups=1, dilation=1):
+    """3x3 convolution with padding"""
+    return torch.nn.Conv2d(in_planes, out_planes, kernel_size=3, stride=stride,
+                           padding=dilation, groups=groups, bias=False, dilation=dilation)
```
Comment (jamesr66a) [inline test/fx/test_gradual_type.py:13]: is `padding=dilation` intended?

File: test/fx/test_gradual_type.py [hunk 7]
```diff
@@ -330,6 +335,76 @@ def forward(self, x: Dyn):
         with self.assertRaises(TypeError):
             tc.type_check()
 
+    def test_type_check_conv2D(self):
```
Comment (jamesr66a) [inline test/fx/test_gradual_type.py:338]: I think a couple things would also make the tests more robust:

1) Can we test that the inferred sizes match with the sizes we actually see when running the module?
2) Can we make the testing more exhaustive? e.g. automate testing over different `kernel_size`, `stride`, `padding`, `dilation`, `groups` parameters and make sure it works across all those parameters

This PR might take a while to get fully tested/correct so it might be a good idea to pull it out from the stack and develop it separately

**Review by jamesr66a — COMMENTED** [review]
Cool, as mentioned inline, there are a lot of degrees of freedom in convolution here and it would be good to try to more thoroughly test across those

File: test/fx/test_gradual_type.py (hunk 6 above)
Comment (migeed-z) [reply test/fx/test_gradual_type.py:13]: yeah, I copied this function from torchvision_models.py and it has padding = dilation for some reason.

File: torch/fx/experimental/graph_gradual_typechecker.py (hunk 3 above)
Comment (migeed-z) [reply torch/fx/experimental/graph_gradual_typechecker.py]: this is for checking if the current node can be expanded to a size 4 tensor so we can check for consistency in the following conditional. I can do this directly in the conditional though instead!

**Comment by migeed-z** [comment]
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29562478).

**Review by jamesr66a — APPROVED** [review]
image

File: torch/fx/experimental/graph_gradual_typechecker.py [hunk 8]
```diff
@@ -207,6 +208,65 @@ def bn2d_inference_rule(n: Node, module_instance):
     else:
         raise TypeError(f'Cannot apply {module_instance} with input type {arg_type} and existing type {n.type} on {n}')
 
+def calculate(d_in, module_instance, index):
```
Comment (jamesr66a) [inline torch/fx/experimental/graph_gradual_typechecker.py:211]: Can this be named something more descriptive?
